Group now groups kw arguments as well

piever opened the pull request:
With this PR, when grouping, also keyword arguments get grouped, provided they are long enough in the first dimension. Fixes #1031 and #962 
New behavior:

```
using Plots
using DataFrames
import RDatasets
gr()
iris = RDatasets.dataset("datasets","iris");
scatter(iris[:PetalLength], iris[:PetalWidth], group = iris[:Species],
layout = 3, marker_z = iris[:SepalLength], markersize = iris[:SepalWidth])
```

![group](https://user-images.githubusercontent.com/6333339/29525465-aa94af80-868a-11e7-82ef-b7d95d8f168a.png)

Also, there was a forgotten method to group, which gives a Dict of indices instead of a column: it is fixed with this PR. Example syntax:

`scatter(iris[:PetalLength], iris[:PetalWidth], group = Dict(:a => collect(1:50), :b => collect(51:150)))`

Issue with this PR: in case of overlaid plots with colorbar, we get superimposed colorbars. We'd need a working keyword to make sure the user can set them to be the same:

`scatter(iris[:PetalLength], iris[:PetalWidth], group = iris[:Species], marker_z = iris[:SepalLength])
`
![groupbad](https://user-images.githubusercontent.com/6333339/29525615-2b5a4134-868b-11e7-9a61-db53548da03b.png)

mkborregaard commented:
Very nice.

mkborregaard commented:
Does it work with StatPlots, and with passing markershapes as a 1-row Matrix?

mkborregaard commented on src/args.jl:
```diff
@@ -854,7 +854,7 @@ end
 function extractGroupArgs{T, V<:AVec{Int}}(idxmap::Dict{T,V}, args...)
     groupLabels = sortedkeys(idxmap)
-    groupIds = VecI[collect(idxmap[k]) for k in groupLabels]
+    groupIds = Vector{Int}[collect(idxmap[k]) for k in groupLabels]
```
Must be a leftover from some old code

mkborregaard commented on src/series.jl:
```diff
@@ -511,10 +511,17 @@ end
             label     --> string(glab)
             idxfilter --> groupby.groupIds[i]
+            for (key,val) in d
+                if key != :group && isa(val, AbstractArray) && size(val,1) >= lengthGroup
```
why `>=`?

piever replied:
You're probably right, I'm changing it to `==`

piever commented:
`markershape` works:
```
scatter(iris[:PetalLength], iris[:PetalWidth], group = iris[:Species],
markersize = iris[:SepalWidth], markershape = [:circle :star :diamond])
```
![group](https://user-images.githubusercontent.com/6333339/29527106-b06d5a88-868f-11e7-9407-1de0be8aab3d.png)

It also works with StatPlots but only with the DataFramesMeta syntax:

`@with iris scatter(:PetalLength, :PetalWidth, group = :Species, layout = 3, marker_z = :SepalLength, markersize = :SepalWidth)`

This one doesn't work:

`scatter(iris, :PetalLength, :PetalWidth, group = :Species, layout = 3, marker_z = :SepalLength, markersize = :SepalWidth)`

But occasionally the StatPlots syntax with the dots behaves a bit strangely, I'm not sure what it is.

piever commented on src/series.jl:
```diff
@@ -511,10 +511,17 @@ end
             label     --> string(glab)
             idxfilter --> groupby.groupIds[i]
+            for (key,val) in d
+                if key != :group && isa(val, AbstractArray) && size(val,1) >= lengthGroup
```
It is possible to give a group dictionary with less elements than the actual data:
`plot(rand(10), group = Dict(:a => [1,3], :b = [2,4]), ms = rand(10))`
In that case I believe we still want to group the markersize.

mkborregaard replied:
I think I disagree here, it seems like it could cause more trouble than benefit.

piever commented:
Maybe there actually is one way to get it to work with StatPlots: I could try and do it [here](https://github.com/JuliaPlots/Plots.jl/blob/b585a0ef61b1af1308cf7ebf0f746c42b39d3869/src/args.jl#L864) instead and then it would also work with the StatPlots syntax. I'll give it a shot and make a second attempt at a PR.

piever commented:
And actually there's one extra corner case I hadn't taken into account: also tuples of vectors of the correct length should be grouped, for example, when passing asymmetric ribbons:
`plot(rand(10), ribbon = (ones(10), 0.5*ones(10)))`

Given that we're fixing `group`, I think we take the opportunity to also fix `:groupedbar` (see [#44](https://github.com/JuliaPlots/StatPlots.jl/issues/44) and [#56](https://github.com/JuliaPlots/StatPlots.jl/issues/56) ). The change would have to happen in the same section I'm modifying now.

piever commented:
Added `tuple` support (for asymmetric ribbons) and addressed your concern. I think all is well now considering the new syntax in DataFrames.
One last thing: I also wanted to add a new method to group, for grouping on more than one column, in the same way as I do with `groupapply`. For example:

```
using DataFrames
import RDatasets
using StatPlots
gr()
school = RDatasets.dataset("mlmRev","Hsb82")
@given school density(:MAch, group = (:Sx, :Sector))
```

would split in 4 groups depending on all the combinations of `Male` `Female` and `Public` `Private` and the legends would be

```
Male Public
Male Private
Female Public
Female Private
```

is something that I end up doing pretty often with my data (for example when having 3 separate experimental groups and splitting by treatment vs no treatment). It's pretty easy to implement, if you're also happy to have this feature I can add it here.

mkborregaard commented:
yes, sounds great :-)

piever commented:
Feature added:

`@given school density(:MAch, group = (:Sx, :Sector))
`

![density](https://user-images.githubusercontent.com/6333339/29752633-0cfed0d2-8b5a-11e7-84ed-fe83753d5c3f.png)

piever commented:
I'll merge later today if everybody is happy with the changes.

mkborregaard commented:
:tada:

mkborregaard commented:
I feel like the example https://github.com/JuliaPlots/Plots.jl/pull/1033#issuecomment-325214217 should be added to the StatPlots readme. It's really cool.
